src/EventStore.Client/GossipChannelSelector.cs

```diff
@@ -0,0 +1,213 @@
+			_settings = settings;
+			_serverCapabilities = serverCapabilities;
+			_cancellationToken = cancellationToken;
+			_grpcChannelsCache = new Dictionary<EndPoint, ChannelBase>();
+			_serverCapabilitiesCache = new Dictionary<EndPoint, ServerCapabilities>();
+			_channelInfoChannel = System.Threading.Channels.Channel.CreateUnbounded<ChannelInfo>();
+			_controlChannel = System.Threading.Channels.Channel.CreateUnbounded<ControlMessage>();
+			_log = settings.LoggerFactory?.CreateLogger<GossipChannelSelector>() ??
+			       new NullLogger<GossipChannelSelector>();
+
+			foreach (var endPoint in settings.ConnectivitySettings.GossipSeeds) {
+				_grpcChannelsCache.Add(endPoint, ChannelFactory.CreateChannel(settings, endPoint));
+			}
+
+			_controlChannel.Writer.TryWrite(ControlMessage.Discover.Instance);
+
+			_ = Task.Run(async () => {
+				var comparer = settings.ConnectivitySettings.NodePreference switch {
+					NodePreference.Leader => NodePreferenceComparers.Leader,
+					NodePreference.Follower => NodePreferenceComparers.Follower,
+					NodePreference.ReadOnlyReplica => NodePreferenceComparers.ReadOnlyReplica,
+					_ => NodePreferenceComparers.Random
+				};
+
+				Loop:
+				while (await _controlChannel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false)) {
+					var message = await _controlChannel.Reader.ReadAsync(cancellationToken).ConfigureAwait(false);
+
+					if (message is ControlMessage.SetEndPoint setEndPoint) {
+						await UseEndPoint(setEndPoint.EndPoint).ConfigureAwait(false);
+					} else {
+						for (var attempt = 1; attempt <= settings.ConnectivitySettings.MaxDiscoverAttempts; attempt++) {
+							var (success, endPoint, nodes) = await TryDiscover(attempt).ConfigureAwait(false);
+
+							if (!success) {
+								continue;
+							}
+
+							await UseEndPoint(endPoint!).ConfigureAwait(false);
+
+							foreach (var endpoint in _grpcChannelsCache.Keys.ToArray()) {
+								if (nodes!.Any(node =>
+									    node.EndPoint.Host == endpoint.GetHost() &&
+									    node.EndPoint.Port == endpoint.GetPort())) {
+									continue;
+								}
+
+								if (!_grpcChannelsCache.TryGetValue(endpoint, out var channel)) {
+									continue;
+								}
+
+								_grpcChannelsCache.Remove(endpoint);
+								await channel.DisposeAsync().ConfigureAwait(false);
+							}
+
+							goto Loop;
+						}
+
+						_log.LogError("Failed to discover candidate in {maxDiscoverAttempts} attempts.",
+							settings.ConnectivitySettings.MaxDiscoverAttempts);
+
+						_channelInfoChannel.Writer.TryComplete(
+							new DiscoveryException(settings.ConnectivitySettings.MaxDiscoverAttempts));
+						_controlChannel.Writer.TryComplete();
+					}
+				}
+
+				async ValueTask<(bool, EndPoint?, ClusterMessages.MemberInfo[]?)> TryDiscover(int attempt) {
+					foreach (var channel in _grpcChannelsCache.Values.ToArray().OrderBy(_ => Guid.NewGuid())) {
```

Comment:
it's possible the OrderBy method would be expecting the keyselector to give a consistent ordering (presumably it could call the key selector multiple times for the same element and be confused if the element has 'moved')

Reply:
(also not sure we need the toarray here)

Reply:
It shouldn't work but it does: https://code.sololearn.com/cyeifUx43RBd

Reply:
We do need it, `nodes` is returned and will be enumerated more than once.